### ISSUE TITLE
Deserialization grammar misses `tree_node_child`

### ISSUE THREAD

**Issue by dmgolembiowski** [dmgolembiowski/edgemorph#12]
# TL;DR 
[`edm/src/common/deserialize_ast.pest`](https://github.com/dmgolembiowski/edgemorph/blob/master/edm/src/common/deserialize_ast.pest#L225) isn't capturing `tree_node_child`, and it needs to. 

## Overview
The simplest of AST deserialization examples yields a surprisingly large node tree. I've observed that EdgeDB's SDL parser will generate a schema in case one is not supplied (a good thing), and this makes it tougher to catch a simple inductive hypothesis. I spent the better part of three hours on [prettier.io](https://prettier.io/playground) and [pest.rs](https://pest.rs/#editor) to find out where the problem lies, but still no luck yet. Hopefully tonight I'll work out where the grammatical flaw lies. 

### SDL module (input) 
```graphql
module this_module_name {
    type User {
        required property name -> str;
    }
}
```
The true SDL abstract syntax tree (output) is too unwieldy to put on the issue, so I prepared a link to [the Gist](https://gist.github.com/dmgolembiowski/b8fad11173ed776b5e2d5fc22fffae80).
The full version is also available at [prettier.io/playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAKgJznAchAEzgAIAdKUmASwIF4BGAFgAYmAmZgNns4A4BmTp0YAaMlGKTiUAIYBbOLQDKYABZw5MseSnE1VADYFsUWgG0du4hmx5CcAMKrDBcVak1a+KHG0T3xAYyAEZwBrQA5ERgQZgy1NAAzhF+AdL2tKgAMlSJMMSeTKyM9Lwc-PxsAKyMom5p7lTwcom0TnBgANZwBDl5ZgCyMpidAK4ADgC6ABQW-g26lJQ2OPhE9QsLhSzsXDy8jACc9GypmwuyCpEDhKMGcAAiHbHxVNApG+cB+kYmbepdHp9GBmLCrexOFwzOZfBpLGArOxESFGT6w86eby+NHohZBULhCKXOAfSy485QDKItYkbbFUrlQT0E5ncls4mRADywQAVh0YAAlOAAM1J8zZuJ+xgQ-w63V6uRBYKRjmcRmhZIl6Ph1Ihatcmq1RrI1DoWNZxst+LCkWJYstDvSREyShgmCoUAA5sQ8phIjBnIkAPpyW73IN2gB8hqk8Lj+RN8cjkwAlJHIxaHbrkfqcQ7MdBsTGHdbCdEXgkoMlM47Kc7soqCnQikwGcwKpVGLwa47Gs1Wu1AQr+kMRhMNeLewEdcqaU2GDsSmV24IanVi1OKfJFBEHNh4nBuXywDB0ABPcYknubzZSv6D+XA0G2GkoggTm-kmcvvUuPOf8kCx8a8APRUtbW3e1QKNOtFGzWlm0XNsKk4KpDlOf9oK1DkIiPfkhVFECsLZO8ZQfIFFWfcEcyhWYN2Itlv2o1U-3ohjHUoICi0ndjN0ocCiUgojeONShYJdN0PW9X1IgAVUSOBMAidNU3TC143jeC30wo0uOEiUBMgORNCgAhqx07CqWBecWyXRk2E4AB2DC2KnJoNAHAFH0o0cximOieIApiVRspDlxQ9Duws4icL3OADwcaAwGweAAAVMAgS9MBgM8oJEqxSNMcjhyVH8aPVAL8vcYLX1zVyqskPTooagTciDbAAEdRiobACDyhqGnEmxRkcJIYBkWBz0vNScQ0mA5rmrS6sCgbdCa+qWpCG0InGzBPTgGB+tW9whtnexQvpcLBEORzOBEDbjqkHCprwITmseyRCtlIcnzO8r30qj6vhq39UQeoG0k4s1C30iG4RgATNA9HLLyOuHtRgU6ypIcH0a+OlWyuxzmDYdcVrxiUcLwk8CLRimtS+4rfuxt8P3ph0Qf+972cA6HgO5nmwK2wk7VhwXcSG113S9H03UiX1lOTNMlZm3GGKW1jyeO9atdWgTRYF3iseYi7CcZRhnOYMWPqp3l8JFOn6a+v6WIqmFxYaDWwd1j2dY9zZ9betWKclySZZkwSFEV1SVdTa2RMk4yeloAAxGQDAU+PeOCOJARgVoIjMSZo7TOPDZvROFDoNOM+4x6c5kPOC6LkulczSvk5rzO0QbpvIhblTIxAEQQEyytEmQUBhgygB3VLhgQCeUHTmeZDPCeR977oYCUcZG6k5A3RGkeNFCAgiF6CbPVGGR9pTiBME0GBqC9ZAQBkUYYAgYeQFUGA5AMAAdWcPARIe8wBwCUIvJoVAABuTQzxvzAIkDeIAPQKWyulW+mhkDCnTpnEAPJEgAA8ABCudt5KG3DkHwuD8G+EISQpQUl7gAEVRgQHgHQ2uI896YAwW-YIwsf7jGljAQBNAAzIF4FbEAoiIAKUAXEcYb9RFwAwbAuAP8uqcLgOlTKS936JAALQ+B6D0H+nVurYCwZ6HBSA8E8JAApOQVBD6YGPs4lhcB2G6O4QQ8awQJEECkUgU4IA3QyEMFJRKxkZBv3UVUH+owFLoBCEvRxBDYEjQAJKmQQDvZKVBxgwAAIKmVdGee4-i4AAF9alAA). Here's a snippet of the cleaner output.

```javascript
<TreeNode 
	id=140142061683664, name=Schema, children=[
      <TreeNodeChild 
        id=None, 
        label='declarations', 
        node=<List id=140141820332544, 
               items=CheckedList[Markup]([
               		<TreeNode 
                             id=140142061684912, 
                             name='ModuleDeclaration', 
                              children=CheckedList[TreeNodeChild]([
                                       <TreeNodeChild 
                                               id=None, ...
```
and the complexity only increases from there.

#### 1) `schema` or `root` -> ...
```idl
schema = ${ SOI ~
    "<" ~ __ ~ "TreeNode" ~ __
    ~ "id" ~ "=" ~ id ~ __
    ~ "name" ~ "=" ~ __ ~ "Schema" ~ __
    ~ "children" ~ __ ~ "=" ~ __ ~"[" ~ __ ~ module ~ __ ~ "]" ~ __ ~ ">" ~ EOI
}
```
_* side note*_:
> where:
>- `SOI` and `EOI` mean _start of input_ and _end of input_, respectively;
>- `X ~ Y` means `X` _followed by_ `Y`;
> - ` __` is a blanket catch-all for undesired tokens in the compound atomic groups 
>     - (things like `thing_x = ${ thing_y ~ __ ~  thing_z })` where `__` is junk in between. It includes:
> ```idl
> __ignore = _{
>       " " 
>    | "," 
>    | NEWLINE
>    | "trimmed=False"
>    | ("at 0x" ~ ASCII_ALPHANUMERIC{12})
>    | ("edb." ~ ( ASCII_ALPHA_LOWER* ~ "." )+ )
> }
>
>__ = _{ __ignore* } // The potential kinds of erroneous syntax is
>                  // not feasible to recreate with the EdgeDB QL Parser
>```
>this ^ is useful for making the deserializer more fault-tolerant to accidental modifications —
ensuring that the following:
> `children=edb.common.checked.CheckedList[edb.common.markup.elements.lang.TreeNodeChild]`, or
> ```javascript
> children = edb.common.checked.CheckedList[ 
>    edb.common.markup.elements.lang.TreeNodeChild
> ]
> ```
> still gets reduced to `children=CheckedList[TreeNodeChild]`.

#### 2) ... -> `module` -> `module_node` -> ...

```idl
// `module` is a special case of the the `tree_node_child`
module = { 
     "[" ~ "<" 
     ~ "TreeNodeChild"
     ~ "id" ~ "=" ~ id
     ~ "label" ~ "=" ~ "'declarations'" 
     ~ "node" ~ "=" ~ module_node 
     ~ ">" ~ "]"
}

// `module_node` is a special case of `checked_list`
module_node = {
    "<" ~ "List" ~ " " 
    ~ "id" ~ "=" ~ id ~ __
    ~ "items" ~ "=" ~ __ ~ "CheckedList"
    ~ "[" ~ __ ~ "Markup" ~ __ ~ "]"
    ~ "(" ~ __ ~ "[" ~ __ ~ module_declaration ~ __ ~ "]" ~ __ ~ ")" ~ __
}
```

#### 3) ... -> `module_declaration`

```idl
module_declaration = {
    "<" ~ __ ~  "TreeNode" ~ __
    ~ "id" ~ __ ~ "=" ~ id ~ __
    ~ "name" ~ __ ~ "=" ~ __ ~ "'ModuleDeclaration'" ~ __ // Beware the internal single-quotes ( ' ... ' )
    ~ "children" ~ __ ~ "=" ~ __ ~ "CheckedList" ~ __   
    ~ "[" ~ __ ~ "TreeNodeChild" ~ __ ~ "]" ~ __       
    ~ "<" ~ __ ~ "label" ~ __ ~ "=" ~ "'name'" ~ __   
    ~ "node" ~ __ ~ "=" ~ __ ~ tree_node ~ __
}
```
#### 4) ... -> `tree_node` -> ...

```idl
tree_node = ${
    "<" ~ __ ~ "TreeNode" ~ __
    ~ "id" ~ __ ~ "=" ~ __ ~ id ~ __
    ~ "name" ~ __ ~ "=" ~ __ ~ name ~ __
    ~ "children" ~ __ ~ "=" ~ __ ~ checked_list ~ __
    ~ ( "brackets" ~ "=" ~ brackets ~ __ )? ~ __
    ~ ">"
}
``` 
where: `id = { "None" | ('0'..'9'){15} }`, `str_value = { !(__forbidden_char)+  }` and `name = { str_value }`.

#### 5) ... -> checked_list -> ...

```idl
checked_list = ${
    __ ~ "CheckedList" ~ "[" ~ ( __ ~ "TreeNodeChild" | __ ~ "Markup" ) ~ __ ~ "]"
    ~ "(" ~ __ ~ "[" ~ ( __ ~ tree_node_child)+ ~ "]" ~ __ ~ ")"
}
```
#### 6) ... -> tree_node_child -> ...

```idl
tree_node_child = ${
    "<" ~ "TreeNodeChild" ~ __
    ~ "id" ~ "=" ~ id ~ __
    ~ "label" ~ "=" ~ label ~ __
    ~ "node"  ~ "=" ~ rhs ~ ">"
}
```
where `rhs` is given by:

```idl
rhs = ${
      list_kind 
    | string_kind 
    | atomic_node
    | true_constant_type_kind
}
```
I'm thinking the error probably lies in either `label`, `rhs`, or both. If any clever people notice something, please share it on this thread. :slightly_smiling_face:

**Comment by dmgolembiowski** [comment]
@nsidnev have you ever worked with a formal grammar before? It may be helpful to study https://pest.rs/book/grammars/syntax.html for future super-powers.
